### ISSUE TITLE
Improve sparsity threshold calculation

### ISSUE THREAD

**Issue by markusbattarbee** [fmihpc/vlasiator#432]
Currently we set the sparsity threshold based on the phase space density, and potentially scale it with particle number density.

Some ideas that have been floated for improving on it:
* threshold based on vcell-generated pressure
* use per-cell per-population pressure instead of number density for threshold scaling

**Comment by markusbattarbee** [comment]
New suggestion:
1. Calculate total density contained in blocks which are >sparsity 
2. Calculate total density contained in buffer blocks (<sparsity)
3. If buffer density is greater than a given quality value (1%?) of total density, decrease the sparsity parameter for this cell
4. If buffer density is less than a given quality value (0.01%?) of total density, increase the sparsity parameter for this cell

**Comment by ykempf** [comment]
Sounds like a good idea, physically motivated is always better than sleeve-shaken.

Fine-tuning this pair of parameters sounds tricky but then results are likely not going to differ by much. A simulation of the resulting effective threshold with 2D and 3D restarts we have lying around would be good to see how the spatial distribution of the effective threshold would look like.

Would you predefine a factor by which the threshold gets changed (x2 or :2, x10 or :10) when it has to change?

**Comment by markusbattarbee** [comment]
Yeah, I think x2 / x0.5 would be a good value in order to make slow, careful changes to the sparsity. This will need significant testing, as the dynamics might prove surprising - and we probably want to have some hard-set minimum and maximum values as well. 

The suggestion of evaluating this method with existing restart files sounds good!
